kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CommitMetadataE2ESuite.scala

```diff
@@ -0,0 +1,106 @@
+import io.delta.kernel.defaults.utils.{TestCommitterUtils, WriteUtilsWithV2Builders}
+import io.delta.kernel.engine.Engine
+import io.delta.kernel.utils.CloseableIterable.emptyIterable
+import io.delta.kernel.utils.CloseableIterator
+
+import org.scalatest.funsuite.AnyFunSuite
+
+class CommitMetadataE2ESuite extends AnyFunSuite
+    with WriteUtilsWithV2Builders
+    with ParquetSuiteBase
+    with TestCommitterUtils {
+
+  private class CapturingCommitter extends Committer {
+    val capturedCommitMetadatas = ListBuffer[CommitMetadata]()
+
+    override def commit(
+        engine: Engine,
+        finalizedActions: CloseableIterator[Row],
+        commitMetadata: CommitMetadata): CommitResponse = {
+      capturedCommitMetadatas.append(commitMetadata)
+
+      committerUsingPutIfAbsent.commit(engine, finalizedActions, commitMetadata)
+    }
+  }
+
+  test("transaction passes domain metadata to committer for add and remove") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      // ===== GIVEN =====
+      val txn0 = TableManager.buildCreateTableTransaction(tablePath, testSchema, "engineInfo")
+        .withTableProperties(Map("delta.feature.domainMetadata" -> "supported").asJava)
+        .build(engine)
+      commitTransaction(txn0, engine, emptyIterable())
+
+      // ===== WHEN (Case 1: Add domain metadata) =====
+      val capturingCommitter = new CapturingCommitter()
+      val snapshot1 =
+        TableManager.loadSnapshot(tablePath).withCommitter(capturingCommitter).build(engine)
+      val txn1 = snapshot1.buildUpdateTableTransaction("engineInfo", Operation.WRITE).build(engine)
+      txn1.addDomainMetadata("test.domain1", """{"key1":"value1"}""")
+      commitTransaction(txn1, engine, emptyIterable())
+
+      // ===== THEN (Case 1: Add domain metadata) =====
+      val addedDomainMetadata =
+        capturingCommitter.capturedCommitMetadatas.head.getNewDomainMetadatas.get(0)
+      assert(addedDomainMetadata.getDomain == "test.domain1")
+      assert(addedDomainMetadata.getConfiguration == """{"key1":"value1"}""")
+      assert(!addedDomainMetadata.isRemoved)
+
+      // ===== WHEN (Case 2: Remove domain metadata) =====
+      val snapshot2 =
+        TableManager.loadSnapshot(tablePath).withCommitter(capturingCommitter).build(engine)
+      val txn2 = snapshot2.buildUpdateTableTransaction("engineInfo", Operation.WRITE).build(engine)
+      txn2.removeDomainMetadata("test.domain1")
+      commitTransaction(txn2, engine, emptyIterable())
+
+      // ===== THEN (Case 2: Remove domain metadata) =====
+      val removedDomainMetadata =
+        capturingCommitter.capturedCommitMetadatas(1).getNewDomainMetadatas.get(0)
+      assert(removedDomainMetadata.getDomain == "test.domain1")
+      assert(removedDomainMetadata.isRemoved)
+    }
+  }
+
+  test("transaction passes no domain metadatas to committer if none added in transaction") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      // ===== GIVEN =====
```

Comment:
can we have existing table with existing DMs? I think we want to check we only return NEW domain metadatas which isn't covered by either of these tests

Reply:
+1